### PR TITLE
FEC-57: Add `platform-limits` model (Part 1)

### PR DESCRIPTION
**Pull request by jaikumar-tj**
### Summary

This is the 1st PR in this series to add the `platform-limits` model. I couldn't find any sdk related to platform-limits for defining typescript types, instead I hardcoded the type based on individual model.

The models are copied from FE repo and few adjustment to `spec` file and `typescript` are done. The model covered in this PR are referred from this below code
- [business-unit](https://github.com/commercetools/merchant-center-frontend/tree/main/packages-shared/test-data/src/platform-limits/business-units)
- [limit](https://github.com/commercetools/merchant-center-frontend/tree/main/packages-shared/test-data/src/platform-limits/limit)
- [limit-with-current](https://github.com/commercetools/merchant-center-frontend/tree/main/packages-shared/test-data/src/platform-limits/limit-with-current)

### CODE REVIEW

File: models/platform-limits/src/business-units/types.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import type { TBuilder } from '@commercetools-test-data/core';
+import type { TLimit } from '../limit';
+
+export type TBusinessUnitLimitsProjection = {
+  maxDivisions: TLimit;
+  maxDepthLimit: TLimit;
+  maxAssociates: TLimit;
+  maxAssociateRoles: TLimit;
+};
```
Comment (jaikumar-tj) [inline models/platform-limits/src/business-units/types.ts:9]: I couldn't find the `type` in any of the repo. I wrote them down based on the schema from graphql explorer
Screenshot 2024-08-28 at 08 44 19

File: models/platform-limits/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Export types
+export * from './business-units/types';
+export * from './limit-with-current/types';
+
+// Export models
+export * as BusinessUnitLimitsProjection from './business-units';
```
Comment (jaikumar-tj) [inline models/platform-limits/src/index.ts:6]: should i change the folder name also reflect the API model?
`business-unit-limits-projection` ? or is it OK to keep it same as in FE test-data?

File: models/platform-limits/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+```bash
+$ pnpm add -D @commercetools-test-data/platform-limits
+```
+
+# Usage
+
+```ts
+
+```
```
Comment (jaikumar-tj) [inline models/platform-limits/README.md:17]: Readme is updated in the [part-3 PR](https://github.com/commercetools/test-data/pull/657)

File: models/platform-limits/src/business-units/builder.spec.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import * as BusinessUnitsPlatformLimits from './index';
+
+describe('building', () => {
+  it(
+    ...createBuilderSpec<
+      TBusinessUnitLimitsProjection,
+      TBusinessUnitLimitsProjection
+    >(
+      'default',
+      BusinessUnitsPlatformLimits.random(),
```
Comment (CarlosCortizasCT) [inline models/platform-limits/src/business-units/builder.spec.ts:14]: Here we're testing the default values of the generator, which does not seem to provide too much value since all properties are initialized with `null`.
Perhaps we can use the `withLimit` preset so we make sure the model is instantiated with some actual data.

File: models/platform-limits/src/business-units/transformers.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+const transformers = {
+  default: Transformer<
+    TBusinessUnitLimitsProjection,
+    TBusinessUnitLimitsProjection
+  >('default', {
+    buildFields: [
+      'maxDivisions',
+      'maxDepthLimit',
+      'maxAssociates',
+      'maxAssociateRoles',
```
Comment (CarlosCortizasCT) [inline models/platform-limits/src/business-units/transformers.ts:16]: **nit**: we could extract this list to a constant in the file and reuse it in the three transformers.

File: models/platform-limits/src/index.ts (hunk 2 above)
Comment (CarlosCortizasCT) [reply models/platform-limits/src/index.ts:6]: I vote for keeping it like it currently is as the nested folder nature already tell us this is a platform limit

File: models/platform-limits/src/limit-with-current/presets/with-warning-exceeded.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import LimitWithCurrent from '../builder';
+
+const withWarningExceeded = () => LimitWithCurrent().limit(10).current(8);
```
Comment (CarlosCortizasCT) [inline models/platform-limits/src/limit-with-current/presets/with-warning-exceeded.ts:3]: Shouldn't `current` be a greater number than `limit`?

File: models/platform-limits/src/limit-with-current/presets/with-exceeded.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import LimitWithCurrent from '../builder';
+
+const withExceeded = () => LimitWithCurrent().limit(10).current(10);
```
Comment (CarlosCortizasCT) [inline models/platform-limits/src/limit-with-current/presets/with-exceeded.ts:3]: Shouldn't `current` be a greater number than `limit`?

File: models/platform-limits/src/limit-with-current/generator.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+import { TLimitWithCurrent } from './types';
+
+/**
+ * limit - Number
+ * current - Number
+ */
+const generator = Generator<TLimitWithCurrent>({
+  fields: {
+    limit: fake((f) => f.number.int(10)),
+    current: fake((f) => f.number.int(10)),
```
Comment (CarlosCortizasCT) [inline models/platform-limits/src/limit-with-current/generator.ts:11]: Shouldn't `current` be a lower number than `limit` in the generator? (default value)

File: models/platform-limits/src/business-units/types.ts (hunk 1 above)
Comment (CarlosCortizasCT) [reply models/platform-limits/src/business-units/types.ts:9]: Fair enough 👍

File: models/platform-limits/src/limit-with-current/presets/with-exceeded.ts (hunk 7 above)
Comment (jaikumar-tj) [reply models/platform-limits/src/limit-with-current/presets/with-exceeded.ts:3]: Based on this [util file](https://github.com/commercetools/merchant-center-frontend/blob/main/packages-shared/experimental-components/src/features/platform-limits/shared/utils/utils.js), if the both value are same, then it already exceeded

File: models/platform-limits/src/limit-with-current/generator.ts (hunk 8 above)
Comment (jaikumar-tj) [reply models/platform-limits/src/limit-with-current/generator.ts:11]: They both can have same number. Refer above comment.

File: models/platform-limits/src/business-units/transformers.ts (hunk 5 above)
Comment (jaikumar-tj) [reply models/platform-limits/src/business-units/transformers.ts]: done https://github.com/commercetools/test-data/pull/655/commits/d01351de3bd1ac841ec8139721c78efb1fbcad99

File: models/platform-limits/src/business-units/builder.spec.ts (hunk 4 above)
Comment (jaikumar-tj) [reply models/platform-limits/src/business-units/builder.spec.ts]: done https://github.com/commercetools/test-data/pull/655/commits/617107cad9f4c095a0f686bf60262289367ca677

File: models/platform-limits/src/limit-with-current/generator.ts (hunk 8 above)
Comment (CarlosCortizasCT) [reply models/platform-limits/src/limit-with-current/generator.ts:11]: Thanks for the explanation.

What I mean is that, if they both have the same number, the limit will be considered exceeded, right?
If that's correct, then the default value for this limit that the model will be returning will be in the exceeded state, where perhaps it makes more sense that the default value is an allowed one.

Would that make sense to you?

File: models/platform-limits/src/limit-with-current/presets/with-warning-exceeded.ts (hunk 6 above)
Comment (CarlosCortizasCT) [reply models/platform-limits/src/limit-with-current/presets/with-warning-exceeded.ts:3]: Thanks for the explanation.
I understand the rule you mention, but readers coming to this file will not know about it.

For maintainability purposes, I think it would be easier to understand if we use a larger value. As an alternative, we can also leave the `8` but add a comment about the 80% rule.

File: models/platform-limits/src/limit-with-current/generator.ts (hunk 8 above)
Comment (jaikumar-tj) [reply models/platform-limits/src/limit-with-current/generator.ts:11]: I was thinking that by keeping the default state with same maximum value can generate all the different possible states (eg: not exceeded, with warning exceeded, exceeded). Since we already have a preset for all the different states, if user prefer only `allowed` one or `exceeded one` then they can choose the corresponding preset. 

Do you think we make default to generate only `not-exceeded` value? or both `not-exceeded` & `with-warning-exceeded` (because both can be considered as `allowed` one?
In other words, do you prefer the current is `int(9)` or `int(7)` or something else?

Comment (CarlosCortizasCT) [reply models/platform-limits/src/limit-with-current/generator.ts:11]: My vote wold be for `int(7)` as the default one and users customizing that value or using a preset if they want something else.

Comment (jaikumar-tj) [reply models/platform-limits/src/limit-with-current/generator.ts]: Done https://github.com/commercetools/test-data/pull/655/commits/aa7ad4e28b70b1fa5e0b1d52d4c63aaa99db10d1

**Review by antoniolodias — APPROVED** [review]

File: models/platform-limits/src/limit-with-current/presets/with-warning-exceeded.ts (hunk 6 above)
Comment (jaikumar-tj) [reply models/platform-limits/src/limit-with-current/presets/with-warning-exceeded.ts:4]: Added comment https://github.com/commercetools/test-data/pull/655/commits/fcd86866594399f9f3a762d5eaa7b027d81aa686

**Review by CarlosCortizasCT — APPROVED** [review]
Thanks! 👍
